examples/console/agent.go

```diff
@@ -0,0 +1,370 @@
+	send := json.NewEncoder(io.MultiWriter(a.Conn(), a.LogWriter))
+	for {
+		m := message{}
+		err = recv.Decode(&m)
+		if err != nil {
+			fmt.Fprintf(a.LogWriter, "error decoding message: %v\n", err)
+			break
+		}
+		err = a.handle(m, send)
+		if err != nil {
+			fmt.Fprintf(a.LogWriter, "error handling message: %v\n", err)
+		}
+	}
+}
+
+func (a *Agent) handle(m message, send *json.Encoder) error {
+	if m.Hello != nil {
+		return a.handleHello(*m.Hello, send)
+	}
+	if m.Open != nil {
+		return a.handleOpen(*m.Open, send)
+	}
+	if m.Update != nil {
+		return a.handleUpdate(*m.Update, send)
+	}
+	if m.Close != nil {
+		return a.handleClose(*m.Close, send)
+	}
+	return nil
+}
+
+func (a *Agent) handleHello(h hello, send *json.Encoder) error {
+	if a.channel != nil {
+		return nil
+	}
+
+	otherEscrowAccountKey, err := keypair.ParseAddress(h.EscrowAccount)
+	if err != nil {
+		return fmt.Errorf("parsing other's escrow account: %w", err)
+	}
+	otherSignerKey, err := keypair.ParseAddress(h.Signer)
+	if err != nil {
+		return fmt.Errorf("parsing other's signer: %v\n", err)
+	}
+	fmt.Fprintf(a.LogWriter, "other's signer: %v\n", otherSignerKey.Address())
+	fmt.Fprintf(a.LogWriter, "other's escrow account: %v\n", otherEscrowAccountKey.Address())
+	escrowAccountSeqNum, err := getSeqNum(a.HorizonClient, a.EscrowAccountKey.Address())
+	if err != nil {
+		return fmt.Errorf("getting sequence number of escrow account: %w", err)
+	}
+	otherEscrowAccountSeqNum, err := getSeqNum(a.HorizonClient, otherEscrowAccountKey.Address())
+	if err != nil {
+		return fmt.Errorf("getting sequence number of other's escrow account: %w", err)
+	}
+	fmt.Fprintf(a.LogWriter, "escrow account seq: %v\n", escrowAccountSeqNum)
+	fmt.Fprintf(a.LogWriter, "other's escrow account seq: %v\n", otherEscrowAccountSeqNum)
+	a.channel = state.NewChannel(state.Config{
+		NetworkPassphrase: a.NetworkPassphrase,
+		MaxOpenExpiry:     openExpiry,
+		Initiator:         a.EscrowAccountKey.Address() > otherEscrowAccountKey.Address(),
+		LocalEscrowAccount: &state.EscrowAccount{
+			Address:        a.EscrowAccountKey,
+			SequenceNumber: escrowAccountSeqNum,
+		},
+		RemoteEscrowAccount: &state.EscrowAccount{
+			Address:        otherEscrowAccountKey,
+			SequenceNumber: otherEscrowAccountSeqNum,
+		},
+		LocalSigner:  a.EscrowAccountSigner,
+		RemoteSigner: otherSignerKey,
+	})
+	return nil
+}
+
+func (a *Agent) handleOpen(openIn state.OpenAgreement, send *json.Encoder) error {
+	open, authorized, err := a.channel.ConfirmOpen(openIn)
+	if err != nil {
+		return fmt.Errorf("confirming open: %w", err)
+	}
+	if authorized {
+		fmt.Fprintf(a.LogWriter, "open authorized\n")
+		if a.channel.IsInitiator() {
+			formationTx, err := a.channel.OpenTx()
+			if err != nil {
+				return fmt.Errorf("building formation tx: %w", err)
+			}
+			err = a.Submitter.SubmitFeeBumpTx(formationTx)
+			if err != nil {
+				return fmt.Errorf("submitting formation tx: %w", err)
+			}
+		}
+	}
+	if !open.Equal(openIn) {
```

Comment:
💭  for the `handleOpen` and `handleUpdate` cases were comparing the input object with the updated one to decide if we should return a message

in reality the participant would send the object one last time for the other party to have the final signatures right? Should we do that here?

Might not matter bc we're going to change to the CAP-40 which would change this code anyways, so maybe add a `TODO` to keep track of?

Reply:
> in reality the participant would send the object one last time for the other party to have the final signatures right? Should we do that here?

You're correct, the participant who signs last needs to send their signatures to the other participant. That's what this code does. It checks if the open received is the same as the open after confirming. If they're the same, no signatures were added to the open, and there's no point sharing it with the other participant since it is identical to the open received. If they're different, signatures must have been added, and the client should send a copy back to the other participant.

Lmk if I'm missing something.

Reply:
👍 nope I misread looks good